File: Java/Tema2_Ejercicio7.java
/*
 * TEMA 2: EXPRESIONES Y OPERADORES - EJERCICIO 7
 * ================================================
 * Escriba un programa que calcule el máquina común divisor (MCD) 
 * usando expresiones y operadores.
 * 
 * EXPLICACIÓN:
 * - El MCD se puede calcular usando el algoritmo de Euclides
 * - MCD(a, b) = MCD(b, a % b) hasta que b = 0
 */

import java.util.Scanner;

public class Tema2_Ejercicio7 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Digite el primer número: ");
        int a = scanner.nextInt();
        
        System.out.print("Digite el segundo número: ");
        int b = scanner.nextInt();
        
        // Valores originales
        int numA = a;
        int numB = b;
        
        // Algoritmo de Euclides
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        
        int mcd = a;
        
        System.out.println("\n=== MÁXIMO COMÚN DIVISOR ===");
        System.out.println("Números: " + numA + ", " + numB);
        System.out.println("MCD: " + mcd);
        
        // Mostrar divisores comunes
        System.out.println("\nDivisores comunes:");
        for (int i = 1; i <= mcd; i++) {
            if (numA % i == 0 && numB % i == 0) {
                System.out.print(i + " ");
            }
        }
        System.out.println();
        
        scanner.close();
    }
}

/*
 * SALIDA ESPERADA:
 * ================
 * Digite el primer número: 48
 * Digite el segundo número: 18
 * 
 * === MÁXIMO COMÚN DIVISOR ===
 * Números: 48, 18
 * MCD: 6
 * 
 * Divisores comunes:
 * 1 2 3 6
 */

File: JS/Tema2_Ejercicio7.js
/*
 * TEMA 2: EXPRESIONES Y OPERADORES - EJERCICIO 7
 * ===============================================
 * Calcular MCD (Máximo Común Divisor) usando algoritmo de Euclides.
 * 
 * EXPLICACIÓN:
 * - Algoritmo de Euclides: MCD(a,b) = MCD(b, a % b)
 * - Caso base: MCD(n, 0) = n
 * - Utiliza operador módulo (%)
 */

function mcd(a, b) {
    while (b !== 0) {
        let temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Versión recursiva
function mcdRecursivo(a, b) {
    return b === 0 ? a : mcdRecursivo(b, a % b);
}

const num1 = 48;
const num2 = 18;

console.log("=== MCD (MÁXIMO COMÚN DIVISOR) ===");
console.log(`MCD(${num1}, ${num2}) = ${mcd(num1, num2)}`);
console.log(`MCD recursivo(${num1}, ${num2}) = ${mcdRecursivo(num1, num2)}`);

console.log(`\nMCD(100, 40) = ${mcd(100, 40)}`);
console.log(`MCD(17, 13) = ${mcd(17, 13)}`);

/*
 * SALIDA ESPERADA:
 * === MCD (MÁXIMO COMÚN DIVISOR) ===
 * MCD(48, 18) = 6
 * MCD recursivo(48, 18) = 6
 * 
 * MCD(100, 40) = 20
 * MCD(17, 13) = 1
 */
